src/processors/extractors/DefaultExtractor.ts

```diff
@@ -0,0 +1,161 @@
+import { getStreamsWithFfprobe } from '../../utils/VideoProber';
+import { Vertex } from '../../entity/Vertex';
+
+const DEFAULT_REGEX = /(?:\.sc\.|\.tc\.|简体|繁體)/;
+
+/**
+ * Default extractor prefer CN/TW subtitle, for audio track use default
+ * For subtitle extraction from subtitle files, action.outputExtname must set.
+ */
+export class DefaultExtractor implements Extractor {
+    public trackIdx: number;
+    public inputPath: string;
+    public streamsInfo: any[];
+    public action: ExtractAction;
+    constructor(public vertex: Vertex) {
+        this.inputPath = null;
+        this.action = vertex.action as ExtractAction;
+    }
+
+    public getInputPath(): string {
+        return this.inputPath;
+    }
+
+    public async extractCMD(): Promise<string[]> {
+        this.findInputAndOutputPath();
+        if (this.action.extractTarget === ExtractTarget.KeepContainer) {
+            return null;
+        }
+        if (this.action.extractFrom === ExtractSource.OtherFiles && extname(this.inputPath) === extname(this.vertex.outputPath)) {
+            // the extension name is equal means we just pick files.
+            return null;
+        }
+        await this.findAllStreams();
+        await this.findTrackId();
+        let mapSelector:string;
+        if (this.trackIdx === -1) {
+            if (this.action.extractTarget === ExtractTarget.AudioStream) {
+                mapSelector = '0:a:0';
+            } else {
+                mapSelector = '0:s:0';
+            }
+        } else {
+            mapSelector = `0:${this.trackIdx}`;
+        }
+        let result = ['-i', this.inputPath, '-map', mapSelector];
+        const isCopy = this.action.extractTarget === ExtractTarget.AudioStream;
+        if (isCopy) {
+            result = result.concat(`-c:a`, 'copy');
+        }
+        return result;
+    }
+
+    private async findAllStreams(): Promise<void> {
+        this.streamsInfo = await getStreamsWithFfprobe(this.inputPath);
+    }
+
+    private findFile(filePaths: string[], type: 'video' | 'audio' | 'subtitle'): string {
+        const files = filePaths.filter(f => {
+            const ext = extname(f);
+            switch(type) {
+                case 'video':
+                    return VIDEO_FILE_EXT.includes(ext);
+                case 'audio':
+                    return AUDIO_FILE_EXT.includes(ext);
+                case 'subtitle':
+                    return SUBTITLE_EXT.includes(ext);
+            }
+        });
+        if (files.length === 1) {
+            return files[0];
+        } else if (files.length > 1) {
+            return files.find(f => {
+                if (this.action.extractRegex) {
+                    const regex = new RegExp(this.action.extractRegex);
+                    return regex.test(f);
+                } else {
+                    return DEFAULT_REGEX.test(f);
+                }
+            });
+        } else {
+            return null;
```

Comment:
Is it safe to return null here?

Reply:
I want to use null to indicate the no file is found. But you may be right, since the inputPath may be null here. I should have documented the inputPath to avoid potential issue.

Reply:
Well, I will review the how implement to see if it's properly handled.

Reply:
I have changed the return type of `getInputPath()` from `string` to `string|null`. and add comment, so the consumer of the API should be aware of the return value can be null.